WIP: Widget Progressbar

luchr opened the pull request:
# Work in Progress

Attempt for a progressbar widget:

[progress.webm](https://user-images.githubusercontent.com/7441996/206906421-a03aa43b-cbe1-4f9d-831f-d8e9ec62d4f2.webm)

Do you have suggestions, wishes and/or remarks?

Above example was generated with 
```python
import time

import TermTk
from TermTk import TTkColor, TTkString

FG_WHITE = TTkColor.fg('#ffffff')
FG_BLACK = TTkColor.fg('#000000')
FG_GREEN = TTkColor.fg('#00ff00')
FG_YELLOW = TTkColor.fg('#ffff00')
FG_RED = TTkColor.fg('#ff0000')
BG_GREEN = TTkColor.bg('#00ff00')
BG_YELLOW = TTkColor.bg('#ffff00')
BG_RED = TTkColor.bg('#ff0000')
BG_DEFAULT = TTkColor.bg('#222222')


def green_yellow_red(bar):
    value = bar.value
    fg = FG_GREEN if value <= 0.3 else FG_YELLOW if value <= 0.6 else FG_RED
    return (None, fg + BG_DEFAULT)

def green_yellow_red_text(bar):
    value = bar.value
    fg = FG_GREEN if value <= 0.3 else FG_YELLOW if value <= 0.6 else FG_RED
    text = (
        TTkString('normal', FG_BLACK + BG_GREEN) if value <= 0.3 else
        TTkString('high', FG_BLACK + BG_YELLOW) if value <= 0.6 else
        TTkString('critical', FG_BLACK + BG_RED))

    return (text, fg + BG_DEFAULT)

def fade_green_red(bar):
    value = bar.value
    red, green = round(value*255), round((1-value)*255)
    fg = f"#{red:02x}{green:02x}00"
    return (None, TTkColor.fg(fg) + BG_DEFAULT)


def create_bar(parent, delay, delta_value, look_and_feel=None):
    bar = TermTk.TTkProgressBar(parent=parent, look_and_feel=look_and_feel)

    timer = TermTk.TTkTimer()

    def timer_cb():
        bar.value += delta_value
        if bar.value < 1.0:
            timer.start(delay)

    timer.timeout.connect(timer_cb)
    timer.start(delay)

time.sleep(3)

TermTk.TTkLog.use_default_file_logging()

root = TermTk.TTk()

win = TermTk.TTkWindow(
    parent=root, pos=(2, 2), size=(40, 10),
    title='Progress', border=True, layout=TermTk.TTkVBoxLayout())

create_bar(win, 0.09, 0.01)
create_bar(win, 0.2, 0.02, green_yellow_red)
create_bar(win, 0.15, 0.01, fade_green_red)
create_bar(win, 0.2, 0.02, green_yellow_red_text)


root.mainloop()
```

ceccopierangiolieugenio commented:
When I implement a new widget I usually check which api are exposed in QT:
https://doc.qt.io/qt-6/qprogressbar.html
I think It looks nice
also the qt version with the percentage at the end looks nice:
![image](https://user-images.githubusercontent.com/8876552/206908336-d545e3ed-12d8-4661-bbe2-3f5a7ffdff39.png)

luchr commented:
OK. I place the (customizable) text at the end.

Because the "blocks" to draw the progress bar are determined by the theme, I can change the standard theme to use

```
 ■■◧
 ■■■
 ■■■◧
 ■■■■
 ■■■■◧
```
instead of the 1/8, 2/8, 3/8, etc. boxes:

```
 █▏
 █▎
 █▍
 █▌
 █▋
 █▊
 █▉
 ██
```
to approximate the QT-image a bit better.

ceccopierangiolieugenio commented on TermTk/TTkWidgets/progressbar.py:
```diff
@@ -0,0 +1,91 @@
+from TermTk.TTkCore.signal    import pyTTkSignal
+from TermTk.TTkWidgets.widget import TTkWidget
+
+def default_look_and_feel(bar):
```
This should be a "@staticmethod" inside TTkProgressBar to avoid to be exported as a standalone api.

ceccopierangiolieugenio commented:
I checked the api,
I'd like to make it more close the the qt ones;
https://doc.qt.io/qt-6/qprogressbar.html
```python
value()
maximum()
minimum()
# slots:
setValue(v)
setMaximum(m)
setMinimum(m)
```

I started using the @property in some templates and widgets when I started my project but I stopped using it because I had some issues at the beginning with complex widgets and also to keep all the apis more consistent (and I am planning to align them in the future to the standard setter/getter pattern) 

about the look and feel,
I always wondered the possibility to allow some widgets to have a "style" attribute but I never formalised it.

The input vars to keep be consistent with the other widgets (and qt names), should be "**maximum**", "**minimum**", "**lookAndFeel**" (or "**style**")

ceccopierangiolieugenio commented on TermTk/TTkWidgets/progressbar.py:
```diff
@@ -0,0 +1,91 @@
+#!/usr/bin/env python3
```
You can add the licence and your name

luchr commented:
Sorry for the delay. 

* I've merge the other PR you made
* The `Allow edits and access to secrets by maintainers` is and has been always checked ... I don't know why you couldn't change ma PR
* I've changed the getter/setter methods according to your comment  https://github.com/ceccopierangiolieugenio/pyTermTk/issues/78#issuecomment-1354759740  ; to the "other" luchr ...   :-)
* I've added `setMinimumMaximum` in order to make it possible to set an new interval (without having the trouble that the separate calls to `setMinimum` and `setMaximum` complain about the minimum being to large or the maximum being to small)
* I've put the (customizable) text at the right end
* Because the text is customizable (dynamically at run-time) there is no chance to get its terminal width in advance, hence a new paramter: `textWidth`; a value of `textWidth=0` means: "no text".
* "no text" can also be enforced by the look and feel by returning `None` for the text.
* The idea with the `■■■◧□□□□` looked ugly (in my eyes); I've kept the ⅛-bars

Please write, if I've missed something and/or if I should change something.

ceccopierangiolieugenio commented on TermTk/TTkWidgets/progressbar.py:
```diff
@@ -0,0 +1,148 @@
+# MIT License
+#
+# Copyright (c) 2022 Eugenio Parodi <ceccopierangiolieugenio AT googlemail DOT com>
+#                    Luchr          <https://github.com/luchr>
```
Feel free to add your name/e-mail

luchr replied:
The way you've written it, fits perfectly.

ceccopierangiolieugenio commented on TermTk/TTkWidgets/progressbar.py:
```diff
@@ -0,0 +1,148 @@
+        self.valueChanged = pyTTkSignal(float)
+        TTkWidget.__init__(self, *args, **kwargs)
+        self.setLookAndFeel(TTkLookAndFeelPBar())
+        self._textWidth = kwargs.get('textWidth', 4)
```
Should textWidth be a property of the lookAndFeel object?
we can add 2 fields:

showText (boolean)
and textWidth

showTtext can be useful to enable/disable dynamically the text
in particular if you want to re-enable it without knowing the textwidth or just to use the last one.

luchr replied:
Nice. More and more is migrating into the look-and-feel.

ceccopierangiolieugenio commented:
added my prototype here:
https://github.com/ceccopierangiolieugenio/pyTermTk/pull/76

luchr commented:
I've just done two minor changes.

ceccopierangiolieugenio commented:
added resizable progressbar in https://github.com/ceccopierangiolieugenio/pyTermTk/pull/75

[Peek 2022-12-20 17-31.webm](https://user-images.githubusercontent.com/8876552/208730523-d7c3cf29-d902-48c9-9137-c99d24a002e7.webm)

serverowner2003 commented:
I tried this code:
```python
import time
import TermTk
from TermTk import TTkColor, TTkString

FG_WHITE = TTkColor.fg('#ffffff')
FG_BLACK = TTkColor.fg('#000000')
FG_GREEN = TTkColor.fg('#00ff00')
FG_YELLOW = TTkColor.fg('#ffff00')
FG_RED = TTkColor.fg('#ff0000')
BG_GREEN = TTkColor.bg('#00ff00')
BG_YELLOW = TTkColor.bg('#ffff00')
BG_RED = TTkColor.bg('#ff0000')
BG_DEFAULT = TTkColor.bg('#222222')


def green_yellow_red(bar):
    value = bar.value
    fg = FG_GREEN if value <= 0.3 else FG_YELLOW if value <= 0.6 else FG_RED
    return (None, fg + BG_DEFAULT)

def green_yellow_red_text(bar):
    value = bar.value
    fg = FG_GREEN if value <= 0.3 else FG_YELLOW if value <= 0.6 else FG_RED
    text = (
        TTkString('normal', FG_BLACK + BG_GREEN) if value <= 0.3 else
        TTkString('high', FG_BLACK + BG_YELLOW) if value <= 0.6 else
        TTkString('critical', FG_BLACK + BG_RED))

    return (text, fg + BG_DEFAULT)

def fade_green_red(bar):
    value = bar.value
    red, green = round(value*255), round((1-value)*255)
    fg = f"#{red:02x}{green:02x}00"
    return (None, TTkColor.fg(fg) + BG_DEFAULT)


def create_bar(parent, delay, delta_value, look_and_feel=None):
    bar = TermTk.TTkProgressBar(parent=parent, look_and_feel=look_and_feel)

    timer = TermTk.TTkTimer()

    def timer_cb():
        bar.value += delta_value
        if bar.value < 1.0:
            timer.start(delay)

    timer.timeout.connect(timer_cb)
    timer.start(delay)

time.sleep(3)

TermTk.TTkLog.use_default_file_logging()

root = TermTk.TTk()

win = TermTk.TTkWindow(
    parent=root, pos=(2, 2), size=(40, 10),
    title='Progress', border=True, layout=TermTk.TTkVBoxLayout())

create_bar(win, 0.09, 0.01)
create_bar(win, 0.2, 0.02, green_yellow_red)
create_bar(win, 0.15, 0.01, fade_green_red)
create_bar(win, 0.2, 0.02, green_yellow_red_text)
root.mainloop()
```

But I got this error:
```
Traceback (most recent call last):
  File "/mnt/My/Home/Projects/server/.exclude/Test/gui/ttk/test15.py", line 62, in <module>
    create_bar(win, 0.09, 0.01)
  File "/mnt/My/Home/Projects/server/.exclude/Test/gui/ttk/test15.py", line 40, in create_bar
    bar = TermTk.TTkProgressBar(parent=parent, look_and_feel=look_and_feel)
AttributeError: module 'TermTk' has no attribute 'TTkProgressBar'

```
and my pyTermTk version is: **(0.37.0a3)**
